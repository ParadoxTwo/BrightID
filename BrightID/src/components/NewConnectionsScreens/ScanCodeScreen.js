// @flow

import * as React from 'react';
import { Alert, StyleSheet, TextInput, View } from 'react-native';
import { connect } from 'react-redux';
import { RNCamera } from 'react-native-camera';
import Spinner from 'react-native-spinkit';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { parseQrData } from './actions/parseQrData';
import { fetchData } from './actions/fetchData';
import emitter from '../../emitter';
import { removeConnectQrData } from '../../actions';
import { setUpWs } from './actions/websocket';

/**
 * Returns whether the string is a valid QR identifier
 * @param {*} qrString
 */
function validQrString(qrString) {
  return qrString.length >= 42;
}

/**
 * Scan code screen of BrightID
 * ==================================================================
 * displays a react-native-camera view
 * after scanning qrcode - the rtc id is set
 *
 */

type State = {
  scanned: boolean,
  connectionAttempts: number,
};

export class ScanCodeScreen extends React.Component<Props, State> {
  textInput: null | TextInput;

  camera: null | RNCamera;

  state = {
    scanned: false,
    connectionAttempts: 0,
  };

  connectionExpired: TimeoutID;

  socket: { close: () => null };

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(removeConnectQrData());
    emitter.on('connectDataReady', this.navigateToPreview);
    emitter.on('connectFailure', this.handleDownloadFailure);
    emitter.on('profileNotReady', this.subscribeToProfileUpload);
  }

  componentWillUnmount() {
    emitter.off('connectDataReady', this.navigateToPreview);
    emitter.off('connectFailure', this.handleDownloadFailure);
    emitter.off('profileNotReady', this.subscribeToProfileUpload);
    clearTimeout(this.connectionExpired);
  }

  handleDownloadFailure = () => {
    this.setState((prev) => ({
      connectionAttempts: prev.connectionAttempts + 1,
    }));
    if (this.state.connectionAttempts > 1) {
      this.navigateToHome();
    } else {
      this.subscribeToProfileUpload();
    }
  };

  handleBarCodeRead = ({ data }) => {
    const { dispatch, navigation } = this.props;

    if (data.startsWith('Recovery_')) {
      navigation.navigate('RecoveringConnection', {
        recoveryRequestCode: data,
      });
    } else if (validQrString(data)) {
      dispatch(parseQrData(data));
      // If the following `fetchdata()` fails, a "connectFailure" will be emitted,
      // triggering a single retry through a websocket notification.
      dispatch(fetchData());
      this.setState({ scanned: true });
      if (this.textInput) this.textInput.blur();
    }
  };

  subscribeToProfileUpload = () => {
    this.socket = this.props.dispatch(setUpWs());
    this.connectionExpired = setTimeout(this.showProfileError, 90000);
  };

  navigateToPreview = () => {
    this.props.navigation.navigate('PreviewConnection');
  };

  showProfileError = () => {
    Alert.alert(
      'Timeout reached',
      "There was a problem downloading the other person's profile. Please try again.",
    );
    this.setState({ scanned: false });
  };

  navigateToHome = () => {
    this.props.navigation.navigate('Home');
  };

  renderCameraOrWave = () => {
    // either camera is showing or webrtc is connecting
    const { scanned } = this.state;
    if (scanned) {
      return (
        <View style={styles.cameraPreview}>
          <Spinner isVisible={true} size={41} type="Wave" color="#4990e2" />
        </View>
      );
    } else {
      return (
        <View style={styles.cameraPreview}>
          <View style={styles.scanTextContainer}>
            <TextInput
              ref={(c) => {
                this.textInput = c;
              }}
              onChangeText={(value) => {
                this.handleBarCodeRead({
                  type: 'text-input',
                  data: value.trim(),
                });
              }}
              style={styles.searchField}
              placeholder="Scan a BrightID code to make a connection"
              autoCapitalize="none"
              autoCorrect={false}
              textContentType="none"
              underlineColorAndroid="transparent"
            />
          </View>
          <RNCamera
            ref={(ref) => {
              this.camera = ref;
            }}
            style={styles.cameraPreview}
            onBarCodeRead={this.handleBarCodeRead}
            type={RNCamera.Constants.Type.back}
            flashMode={RNCamera.Constants.FlashMode.off}
            permissionDialogTitle="Permission to use camera"
            permissionDialogMessage="We need your permission to use your camera phone"
          >
            <Ionicons name="ios-qr-scanner" size={223} color="#F76B1C" />
          </RNCamera>
        </View>
      );
    }
  };

  render() {
    return <View style={styles.container}>{this.renderCameraOrWave()}</View>;
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
  },
  cameraPreview: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
  },
  scanTextContainer: {
    height: 85,
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
    backgroundColor: '#fff',
    zIndex: 100,
  },
  searchField: {
    fontFamily: 'ApexNew-Book',
    fontSize: 16,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    width: '85%',
  },
});

export default connect((state) => state)(ScanCodeScreen);
