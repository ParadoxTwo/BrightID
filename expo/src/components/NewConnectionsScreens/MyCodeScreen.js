// @flow

import * as React from 'react';
import {
  Image,
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Clipboard,
} from 'react-native';
import Svg, { Path } from 'react-native-svg';
import qrcode from 'qrcode';
import { documentDirectory } from 'expo-file-system';
import { connect } from 'react-redux';
import { parseString } from 'xml2js';
import { path } from 'ramda';
// import Spinner from 'react-native-spinkit';
import Material from 'react-native-vector-icons/MaterialCommunityIcons';
import { genQrData } from './actions/genQrData';
import { encryptAndUploadLocalData } from './actions/encryptData';
import { setUpWs } from './actions/websocket';
import emitter from '../../emitter';
import { removeConnectQrData } from '../../actions';

/**
 * My Code screen of BrightID
 *
 * USERA represents this user
 * ==================================================================
 * displays a qrcode
 *
 */

type State = {
  copied: boolean,
  qrsvg:
    | string
    | {
        svg: {
          $: {
            viewBox: string,
          },
        },
      },
};

const COPIED_TIMEOUT = 500;

export class MyCodeScreen extends React.Component<Props, State> {
  connectionExpired: TimeoutID;

  socket: { close: () => null };

  state = {
    qrsvg: '',
    copied: false,
  };

  componentDidMount() {
    // After 2 minutes, connection attempts expire on the server.
    //  Let users start the connection over after one minute so there will be
    //  time for the other user to connect.
    const { dispatch } = this.props;
    this.connectionExpired = setTimeout(this.navigateToHome, 60000);
    emitter.on('connectDataReady', this.navigateToPreview);
    dispatch(removeConnectQrData());
    dispatch(genQrData()).then(() => {
      this.genQrCode();
      this.socket = dispatch(setUpWs());
      dispatch(encryptAndUploadLocalData());
    });
  }

  componentWillUnmount() {
    clearTimeout(this.connectionExpired);
    emitter.off('connectDataReady', this.navigateToPreview);
    if (this.socket) this.socket.close();
  }

  navigateToPreview = () => {
    this.props.navigation.navigate('PreviewConnection');
  };

  navigateToHome = () => {
    this.props.navigation.navigate('Home');
  };

  genQrCode = () => {
    const {
      connectQrData: { qrString },
    } = this.props;
    qrcode.toString(qrString, this.handleQrString);
  };

  handleQrString = (err, qr) => {
    if (err) return console.log(err);
    parseString(qr, this.parseQrString);
  };

  parseQrString = (err, qrsvg) => {
    if (err) return console.log(err);
    this.setState({ qrsvg });
  };

  copyQr = () => {
    const {
      connectQrData: { qrString },
    } = this.props;
    Clipboard.setString(qrString);
    this.setState({ copied: true }, () =>
      setTimeout(() => this.setState({ copied: false }), COPIED_TIMEOUT),
    );
  };

  renderCopyQr = () => (
    <TouchableOpacity style={styles.copyContainer} onPress={this.copyQr}>
      <Material
        size={24}
        name="content-copy"
        color="#333"
        style={{ width: 24, height: 24 }}
      />
      <Text style={styles.copyText}> Copy</Text>
    </TouchableOpacity>
  );

  renderSpinner = () => (
    <View style={styles.qrsvgContainer}>
      {/* <Spinner
        // style={styles.spinner}
        isVisible={true}
        size={47}
        type="9CubeGrid"
        color="#4990e2"
      /> */}
    </View>
  );

  renderQrCode = () => (
    <View style={[styles.qrsvgContainer]}>
      <Svg
        height="212"
        width="212"
        xmlns="http://www.w3.org/2000/svg"
        viewBox={path(['svg', '$', 'viewBox'], this.state.qrsvg)}
        shape-rendering="crispEdges"
      >
        <Path
          fill={this.state.copied ? 'lightblue' : '#fff'}
          d={path(['svg', 'path', '0', '$', 'd'], this.state.qrsvg)}
        />
        <Path
          stroke="#000"
          d={path(['svg', 'path', '1', '$', 'd'], this.state.qrsvg)}
        />
      </Svg>
    </View>
  );

  render() {
    const { photo, name } = this.props;
    const { qrsvg } = this.state;
    return (
      <View style={styles.container}>
        <View style={styles.topHalf}>
          <View style={styles.myCodeInfoContainer}>
            <Text style={styles.myCodeInfoText}>
              To make a new connection, you will share your
            </Text>
            <Text style={styles.myCodeInfoText}>
              name, your photo, your score
            </Text>
          </View>
          <View style={styles.photoContainer}>
            <Image
              source={{
                uri: `file://${documentDirectory}/photos/${photo.filename}`,
              }}
              style={styles.photo}
              resizeMode="cover"
              onError={(e) => {
                console.log(e);
              }}
              accessible={true}
              accessibilityLabel="user photo"
            />
            <Text style={styles.name}>{name}</Text>
          </View>
        </View>
        <View style={styles.bottomHalf}>
          {qrsvg ? this.renderQrCode() : this.renderSpinner()}
          {qrsvg ? this.renderCopyQr() : <View />}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    height: '100%',
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'flex-start',
    flexDirection: 'column',
  },
  topHalf: {
    height: '45%',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  bottomHalf: {
    height: '55%',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  myCodeInfoContainer: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 24,
    marginBottom: 36,
  },
  myCodeInfoText: {
    fontFamily: 'ApexNew-Book',
    fontSize: 16,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    textAlign: 'center',
    color: '#4a4a4a',
  },
  photoContainer: {
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  photo: {
    width: 102,
    height: 102,
    borderRadius: 51,
  },
  name: {
    fontFamily: 'ApexNew-Book',
    marginTop: 12,
    fontSize: 20,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    textAlign: 'left',
    color: '#000000',
    textShadowColor: 'rgba(0, 0, 0, 0.32)',
    textShadowOffset: {
      width: 0,
      height: 2,
    },
    textShadowRadius: 4,
  },
  qrsvgContainer: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  copyContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: 25,
    minWidth: 100,
  },
  copyText: {
    color: '#333',
    fontFamily: 'ApexNew-Book',
  },
});

export default connect((state) => state)(MyCodeScreen);
