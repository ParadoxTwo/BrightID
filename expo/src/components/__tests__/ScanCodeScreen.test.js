import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { ScanCodeScreen } from '../NewConnectionsScreens/ScanCodeScreen';

// you can mock any middlewares here if necessary

const props = {
  navigation: { navigate: jest.fn() },
  dispatch: jest.fn(),
};

describe('Testing ScanCodeScreen', () => {
  it('renders as expected', () => {
    const wrapper = renderer.create(<ScanCodeScreen {...props} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
