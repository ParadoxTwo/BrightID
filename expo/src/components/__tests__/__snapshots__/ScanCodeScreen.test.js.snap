// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ScanCodeScreen renders as expected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "height": 85,
          "justifyContent": "center",
          "padding": 10,
          "width": "100%",
          "zIndex": 100,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        placeholder="Scan a BrightID code to make a connection"
        rejectResponderTermination={true}
        style={
          Object {
            "fontFamily": "ApexNew-Book",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "letterSpacing": 0,
            "width": "85%",
          }
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
      />
    </View>
  </View>
</View>
`;
